### PR TITLE
Reconfix integration.

### PR DESCRIPTION
**Pull request by abrodersen**
Initial work to use reconfix for the `local configure IMAGE` command. Currently uses a single schema to read and write configuration, but this will be expanded after some tests can be created.

### CODE REVIEW

**Comment by abrodersen** [comment]
@pimterry When you have a chance, could you take a look at this PR?

File: extras/schemas/resin-os.json [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+                                            },
+                                            "ipv6": {
+                                                "addr-gen-mode": "stable-privacy",
+                                                "method": "auto"
+                                            }
+                                        }
+                                    },
+                                    {
+                                        "value": false,
+                                        "template": {}
```
Comment (jviotti) [inline extras/schemas/resin-os.json]: It'd be nice to test this on an actual resin device. I want to make sure NetworkManager will not get confused by an empty file.

File: gulpfile.coffee [hunk 2]
```diff
@@ -25,6 +27,10 @@ gulp.task 'coffee', ->
 		.pipe(coffee(bare: true, header: true))
 		.pipe(gulp.dest(OPTIONS.directories.build))
 
+gulp.task 'schemas', ->
+	gulp.src(OPTIONS.files.schemas, { cwd: 'extras' })
```
Comment (jviotti) [inline gulpfile.coffee]: Does this mean we're just copying the schemas to another directory? If so, could we put them on the final directory from the start?

File: lib/actions/local/configure.coffee [hunk 3]
```diff
@@ -95,100 +67,65 @@ inquirerOptions = (data) -> [
 	}
 ]
 
+joinPath = (path) -> '/' + path.join('/')
```
Comment (jviotti) [inline lib/actions/local/configure.coffee]: Keep in mind that Windows paths don't start with a slash.

Comment (jviotti) [reply lib/actions/local/configure.coffee]: Ah, is this Reconfix specific? If so, can we make that clearer, or make the leading slash optional?

File: lib/actions/local/configure.coffee [hunk 4]
```diff
@@ -95,100 +67,65 @@ inquirerOptions = (data) -> [
...
+	imagefs = require('resin-image-fs')
+
+	reconfix = new Reconfix({
+		read: (partition, path) ->
+			imagefs.readFile
+				image: target
+				partition: partition
+				path: joinPath(path)
+			.catch (err) ->
+				if err?.code == 'NOENT'
```
Comment (jviotti) [inline lib/actions/local/configure.coffee]: Shouldn't this be `ENOENT`?

Comment (jviotti) [reply lib/actions/local/configure.coffee]: Why is it necessary to check if `err` exists? It should always be present on a catch block, right?

File: lib/actions/local/configure.coffee [hunk 5]
```diff
@@ -95,100 +67,65 @@ inquirerOptions = (data) -> [
...
+			.catch (err) -> {}
+		write: (partition, path, data) ->
+			imagefs.writeFile
+				image: target,
+				partition: partition,
+				path: joinPath(path)
+			, data.toString('utf-8')
+	})
+	readFileAsync(path.join(__dirname, '../../schemas/resin-os.json'))
+	.then (content) -> JSON.parse(content)
```
Comment (jviotti) [inline lib/actions/local/configure.coffee]: I think you can just do `.then(JSON.parse)`

File: lib/actions/local/configure.coffee [hunk 6]
```diff
@@ -95,100 +67,65 @@ inquirerOptions = (data) -> [
...
+	.then (json) ->
```
Comment (jviotti) [inline lib/actions/local/configure.coffee]: Not critical, but I guess you could write this as:

```
.then(reconfix.loadSchema)
.return(reconfix)
```

File: lib/actions/local/configure.coffee [hunk 7]
```diff
@@ -95,100 +67,65 @@ inquirerOptions = (data) -> [
...
+writeConfiguration = (config, target, type) ->
+	load(target, type)
+	.then (reconfix) -> reconfix.writeValues(config)
+
+
 getConfiguration = (data) ->
-	_ = require('lodash')
 	inquirer = require('inquirer')
 
+	console.log('read configuration: ' + JSON.stringify(data, '  '))
```
Comment (jviotti) [inline lib/actions/local/configure.coffee]: I guess this can be removed?

File: lib/actions/local/configure.coffee [hunk 8]
```diff
@@ -95,100 +67,65 @@ inquirerOptions = (data) -> [
...
+
 	# `persistentLogging` can be `undefined`, so we want
 	# to make sure that case defaults to `false`
 	data = _.assign data,
 		persistentLogging: data.persistentLogging or false
 
 	inquirer.prompt(inquirerOptions(data))
+	.then (answers) ->
+		keys = _.filter(_.keys(answers), (key) -> key != 'advancedSettings')
```
Comment (jviotti) [inline lib/actions/local/configure.coffee]: I think this could be simplified to `_.keys(_.omit(answers, [ 'advancedSettings' ]))`

File: lib/actions/local/configure.coffee [hunk 9]
```diff
@@ -95,100 +67,65 @@ inquirerOptions = (data) -> [
...
+		return _.pick(answers, keys)
```
Comment (jviotti) [inline lib/actions/local/configure.coffee]: Hm, wait, so you're basically returning the original object, just without the `advancedSettings` property, right? If so `_.omit(answers, [ 'advancedSettings' ])` should do the trick

File: lib/actions/local/configure.coffee [hunk 10]
```diff
@@ -95,100 +67,65 @@ inquirerOptions = (data) -> [
 	}
 ]
 
+joinPath = (path) -> '/' + path.join('/')
+
+load = (target, type) ->
+	{ Reconfix } = require('reconfix-preview')
```
Comment (jviotti) [inline lib/actions/local/configure.coffee]: Lets push this package as `reconfix`.

Comment (jviotti) [inline lib/actions/local/configure.coffee]: Is there any reason why the `reconfix-preview` package exports a `Reconfix` property with the actual class? Can we export the class directly?

File: lib/actions/local/configure.coffee [hunk 11]
```diff
@@ -206,22 +143,20 @@ module.exports =
...
-				return if not isMounted
-				umountAsync(params.target)
-		.then (configurationSchema) ->
+		isMountedAsync(params.target).then (isMounted) ->
+			return if not isMounted
+			umountAsync(params.target)
+		.then ->
 			denymount params.target, (cb) ->
-				reconfix.readConfiguration(configurationSchema, params.target)
+				readConfiguration(params.target, '')
```
Comment (jviotti) [inline lib/actions/local/configure.coffee]: What does an empty type mean here?

File: lib/actions/local/configure.coffee [hunk 12]
```diff
@@ -206,22 +143,20 @@ module.exports =
...
 				.then(getConfiguration)
 				.then (answers) ->
-					if not answers.hostname
-						removeHostname(configurationSchema)
-					reconfix.writeConfiguration(configurationSchema, answers, params.target)
+					console.log('answers: ' + JSON.stringify(answers, '  '))
```
Comment (jviotti) [inline lib/actions/local/configure.coffee]: I guess this can be removed

File: lib/actions/local/configure.coffee [hunk 13]
```diff
@@ -206,22 +143,20 @@ module.exports =
...
+					return answers
+				.then (answers) ->
+					writeConfiguration(answers, params.target, '')
```
Comment (jviotti) [inline lib/actions/local/configure.coffee]: Same here. Why do we use an empty path?

File: lib/utils/helpers.coffee [hunk 14]
```diff
@@ -0,0 +1,134 @@
+###
```
Comment (jviotti) [inline lib/utils/helpers.coffee]: Looks like this file is not being used anywhere.

**Comment by pimterry** [comment]
Sorry @abrodersen I know this is waiting on me. It is in my list, just busy with lots of multicontainer stuff at the mo! I'll definitely get to it this week, hopefully tomorrow.

File: extras/schemas/resin-os.json [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+                    },
+                    "deviceType": {
+                        "type": ["string"],
+                        "mapping": ["/deviceType"]
+                    },
+                    "userId": {
+                        "type": ["string"],
+                        "mapping": ["/userId"]
+                    },
+                    "appUpdagePollInterval": {
```
Comment (lekkas) [inline extras/schemas/resin-os.json]: It would be nice to use the "description" attribute to add more context and additional information (e.g. here we could add more info about the value that's expected here (is it seconds, milliseconds?)

File: lib/actions/local/configure.coffee [hunk 16]
```diff
@@ -95,100 +67,65 @@ inquirerOptions = (data) -> [
 	}
 ]
 
+joinPath = (path) -> '/' + path.join('/')
+
+load = (target, type) ->
```
Comment (lekkas) [inline lib/actions/local/configure.coffee]: Can we rename this function to something a bit more descriptive (e.g. `loadReconfix`, or something else)

File: lib/actions/local/configure.coffee (hunk 10 above)
Comment (lekkas) [reply lib/actions/local/configure.coffee]: Related, where would be the best place to look for Reconfix documentation?

File: gulpfile.coffee (hunk 2 above)
Comment (abrodersen) [reply gulpfile.coffee]: The files are copied to the `build` directory, which is not in source control.

File: lib/actions/local/configure.coffee (hunk 3 above)
Comment (abrodersen) [reply lib/actions/local/configure.coffee]: These paths are used by `resin-image-fs`, which means they are always UNIX style.

File: lib/actions/local/configure.coffee (hunk 4 above)
Comment (abrodersen) [reply lib/actions/local/configure.coffee]: `NOENT` is what the error type that `resin-image-fs` emits. It would be best to standardize (or document) the error interface of `resin-image-fs`.

File: lib/actions/local/configure.coffee [hunk 17]
```diff
@@ -95,100 +67,65 @@ inquirerOptions = (data) -> [
...
+			.then (content) -> Buffer.from(content, 'utf-8')
+			.catch (err) -> {}
+		write: (partition, path, data) ->
+			imagefs.writeFile
+				image: target,
+				partition: partition,
+				path: joinPath(path)
+			, data.toString('utf-8')
+	})
+	readFileAsync(path.join(__dirname, '../../schemas/resin-os.json'))
```
[the remark this answers is not in the document]
Comment (abrodersen) [reply lib/actions/local/configure.coffee]: I didn't know this! Still, having `load` return an asynchronous `Promise` makes calling it cleaner.

File: lib/actions/local/configure.coffee (hunk 6 above)
Comment (abrodersen) [reply lib/actions/local/configure.coffee]: I think that looks better.

File: lib/actions/local/configure.coffee (hunk 3 above)
Comment (pimterry) [reply lib/actions/local/configure.coffee]: In that case, it'd be good to add a comment clarifying this here

File: lib/actions/local/configure.coffee (hunk 9 above)
Comment (abrodersen) [reply lib/actions/local/configure.coffee]: Yes, this is what I was looking for.

File: lib/actions/local/configure.coffee (hunk 10 above)
Comment (abrodersen) [reply lib/actions/local/configure.coffee]: Yes, we can. When I first wrote the export, I thought the module had to be a map.

File: lib/utils/helpers.coffee [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+	resin = require('resin-sdk-preconfigured')
+
+	# Attempt to read manifest from the first
+	# partition, but fallback to the API if
+	# we encounter any errors along the way.
+	imagefs.read
+		image: image
+		partition: 1
+		path: '/device-type.json'
+	.then(rindle.extractAsync)
```
[the remark this answers is not in the document]
Comment (abrodersen) [reply lib/utils/helpers.coffee]: @pimterry There is a `readFile` method that simply returns a `Promise` with a string value containing the contents. However, when I try to call this method, I get TypeScript errors saying it does not exist. Where are the TypeScript definitions for `resin-image-fs`?

[the remark this answers is not in the document]
Comment (pimterry) [reply lib/utils/helpers.coffee]: That's because resin-image-fs itself doesn't have any types, it's just got the adhoc ones that are in this codebase, which only covers the stuff we're already using, and doesn't know anything about different versions of resin-image-fs. The file you're looking for is here: https://github.com/resin-io/resin-cli/blob/master/typings/resin-image-fs.d.ts

Don't know how familiar with TS you are, let me know you need a hand.

File: extras/schemas/resin-os.json (hunk 15 above)
Comment (abrodersen) [reply extras/schemas/resin-os.json]: In the current implementation, there is no description field. The schema format is going to change significantly in the next version.

Comment (pimterry) [inline extras/schemas/resin-os.json]: There's a typo in the schema here: `appUpdagePollInterval`

File: .travis.yml [hunk 19]
```diff
@@ -6,6 +6,8 @@ node_js:
   - "6"
 before_install:
 - npm -g install npm@4
+- curl https://sh.rustup.rs -sSf | sh -s -- -y
+- source $HOME/.cargo/env
```
Comment (pimterry) [inline .travis.yml]: This is going to be a problem. End users `npm install` the CLI on their own machines, often Windows machines, and they're not going to have Rust & Cargo installed. This (and the travis output) suggests that's not going to work, is that right?

I think we did discuss this briefly before, sorry I didn't call it out properly on the previous review. There's two options I think: hold this PR until we have CLI bundling completed (so users don't ever run `npm install` themselves), or set up node pre-gyp to precompile and publish reconfix as a binaries that users can use immediately.

The 2nd is probably more practical, if you've got time to set it up, and we'll want it if we want to use reconfix in any other node projects anyway, so we don't have to add a rust build env to every other project. We already do this for other native CLI packages like node-ext2fs. https://github.com/resin-io/node-ext2fs/pull/11 + https://github.com/resin-io/node-ext2fs/pull/13 are a full setup example I think, which should take you most of the way.

To get this working, you need:
* Versionbot set up (we should do this regardless)
* A CI build for each target platform/node version (see node-ext2fs)
* Enough Node configuration to build the project (we should have this already)
* A bit of npm boilerplate for node-pre-gyp & node-pre-gyp-release (see node-ext2fs)

The end goal is that on every tagged commit to master (from versionbot), a binary for every target platform & node version is built, and published to github releases. On every install of the package (including from CLI users), we look in github releases for a prebuilt binary, and use that immediately if there's a compatible one available (or try to build from scratch if not).

This is a non-trivial chunk of work, might be worth more discussion, feel free to ping me. @jviotti might have opinions on this too.

Comment (jviotti) [reply .travis.yml]: Doing prebuilds is definitely an option (we usually favor `prebuild` rather than `node-pre-gyp` given the latter does some weird dynamic stuff that doesn't play nice with PKG) but proper PKG integration is something that would need to do anyways, right? How far are we with that? I remember seeing that you got this working at this point.

@abrodersen I can help you out setting this up for the Node.js bindings, but we should also do the PKG thing asap here.

File: lib/actions/local/configure.coffee [hunk 20]
```diff
@@ -95,8 +66,48 @@ inquirerOptions = (data) -> [
...
+		read: (partition, path) ->
+			imagefs.readFile
+				image: target
+				partition: partition
+				path: joinPath(path)
+			.catch (err) ->
+				if err.code == 'NOENT'
+					return Buffer.alloc(0)
+				else
+					throw err
```
Comment (jviotti) [inline lib/actions/local/configure.coffee]: Why do we throw an error here if we're going to swallow it two lines later?

File: lib/actions/local/configure.coffee [hunk 21]
```diff
@@ -95,8 +66,48 @@ inquirerOptions = (data) -> [
...
+			.then (content) -> Buffer.from(content, 'utf-8')
+			.catch (err) -> {}
```
Comment (jviotti) [inline lib/actions/local/configure.coffee]: I don't think we should be swallowing all errors here.

File: lib/actions/local/configure.coffee [hunk 22]
```diff
@@ -95,8 +66,48 @@ inquirerOptions = (data) -> [
...
+				image: target,
+				partition: partition,
+				path: joinPath(path)
+			, data.toString('utf-8')
+	})
+	schema = require('../../../extras/schemas/resin-os.json')
+	reconfix.loadSchema(schema)
+	Promise.resolve(reconfix)
+
+readConfiguration = (target) ->
```
Comment (jviotti) [inline lib/actions/local/configure.coffee]: These 2 functions are only used in one place as far as I can see, so maybe worth inlining them?

File: lib/actions/local/configure.coffee [hunk 23]
```diff
@@ -95,8 +66,48 @@ inquirerOptions = (data) -> [
 	}
 ]
 
+# Since these paths are used by `resin-image-fs`, we can safely
+# assume they are always UNIX style.
+joinPath = (path) -> '/' + path.join('/')
+
+loadReconfix = (target) ->
+	{ Reconfix } = require('reconfix-preview')
```
Comment (jviotti) [inline lib/actions/local/configure.coffee]: Did we update the npm module to just `reconfix`?

File: package.json [hunk 24]
```diff
@@ -129,10 +131,10 @@
     "resin-device-init": "^4.0.0",
     "resin-docker-build": "^0.4.0",
     "resin-doodles": "0.0.1",
-    "resin-image-fs": "^2.3.0",
+    "resin-image-fs": "^4.1.1",
     "resin-image-manager": "^5.0.0",
     "resin-preload": "^5.0.0",
-    "resin-sdk": "^7.0.0",
+    "resin-sdk": "^7.2.4",
```
Comment (jviotti) [inline package.json]: Was this update necessary for this PR? Can it be done on a separate PR?

File: lib/actions/local/configure.coffee [hunk 25]
```diff
@@ -95,8 +66,48 @@ inquirerOptions = (data) -> [
 	}
 ]
 
+# Since these paths are used by `resin-image-fs`, we can safely
+# assume they are always UNIX style.
+joinPath = (path) -> '/' + path.join('/')
```
Comment (jviotti) [inline lib/actions/local/configure.coffee]: Manually joining strings look fragile. Node.js provides `path.join()` in the `path` built-in module (`require('path')`). If we always want to use UNIX-style, then you can use `path.posix.join()`.

Comment (jviotti) [reply lib/actions/local/configure.coffee]: Maybe something like this?

```js
path = require('path')
joinPath = (fragments) -> path.posix.join(path.posix.sep, ...fragments)
```

File: lib/actions/local/configure.coffee (hunk 23 above)
Comment (jviotti) [inline lib/actions/local/configure.coffee]: Can we export `Reconfix` directly, rather than exporting an object containing a `Reconfix` property? I should ideally be able to do `Reconfix = require('reconfix')`

File: package.json [hunk 26]
```diff
@@ -13,7 +13,8 @@
     "bin/",
     "build/",
     "doc/",
-    "lib/"
+    "lib/",
+    "extras/"
```
Comment (jviotti) [inline package.json]: Where does this directory come from?

File: .travis.yml (hunk 19 above)
Comment (pimterry) [reply .travis.yml]: We have bundled packages published with all releases, but they don't work perfectly I think, and they're not well tested. It's in the readme, but marked as experimental. Needs finishing, but I've only really got time for quick & essential bug fixes on the CLI right now.

**Comment by shaunmulligan** [comment]
@jviotti and @zrzka I assume this PR can be closed as we will integrate the new reconfix later?

**Comment by zrzka** [comment]
@shaunmulligan yes, it's obsolote.
